### PR TITLE
fix(datadog_agent)!: align `ddtags` parsing with DD logs intake

### PR DESCRIPTION
**Pull request by neuronull**
Recently we added a `parse_ddtags` config setting, but the parsing behavior parsed the tags into an object. This updates the parsing behavior to align with the DD logs products.

Additionally, the `datadog_logs` sink was adapted to re-construct the `tags` into the unified comma delimited string as is expected by the logs intake.

### CODE REVIEW

File: changelog.d/dd_agent_parse_ddtags_format.breaking.md
Comment (bruceg) [inline changelog.d/dd_agent_parse_ddtags_format.breaking.md:1]:
```suggestion
Previously the `datadog_agent` setting `parse_ddtags` parsed the tag string into an Object. It is now parsed into an Array of `key:value` strings, which matches the  behavior of the Datadog logs backend intake.
```

**Review by bruceg — APPROVED** [review]

**Review by pront — APPROVED** [review]

**Review by estherk15 — APPROVED** [review]

**Comment by neuronull** [comment]
Note- there is a change needed to the DD logs sink to re-construct the comma delimited string

**Comment by neuronull** [comment]
Also needs a breaking change note in the upgrade guide

**Review by pront — APPROVED** [review]
Approving after re-reviewing

File: src/sinks/datadog/logs/sink.rs
Comment (bruceg) [inline src/sinks/datadog/logs/sink.rs:118]: I think you meant to say this:
```suggestion
                    if let Some(bytes) = tag_kv.as_bytes() {
                        let tag_kv_str = String::from_utf8_lossy(bytes);
```

File: src/sinks/datadog/logs/sink.rs
Comment (bruceg) [inline src/sinks/datadog/logs/sink.rs:122]: Since you're enumerating here, you can use `idx > 0` to push the comma before the string…

File: src/sinks/datadog/logs/sink.rs
Comment (bruceg) [inline src/sinks/datadog/logs/sink.rs:112]: … and then you don't need to remember `n_tags` and can use:
```suggestion
            if !tags_arr.is_empty() {
```

File: src/sinks/datadog/logs/sink.rs
Comment (bruceg) [inline src/sinks/datadog/logs/sink.rs:131]: Is there not a macro to produce a target path from a constant? If not, you should be able to pre-parse this and store it somewhere else rather than parsing on every event.

File: website/content/en/highlights/2024-03-26-0-37-0-upgrade-guide.md
Comment (bruceg) [inline website/content/en/highlights/2024-03-26-0-37-0-upgrade-guide.md:1]: Is this an extraneous file? The filename says `0-37-0` but the text says `0.38.0`

**Review by bruceg — COMMENTED** [review]
I'm a little uncomfortable with the two changes being part of the same PR, let alone the same changelog. Yes, they are related changes but not really interdependent. Given the other issues I identified below, I think it may be best to split off the sink changes and resubmit after the already-approved changes merge.

**Comment by neuronull** [comment]
> I'm a little uncomfortable with the two changes being part of the same PR, let alone the same changelog. Yes, they are related changes but not really interdependent. Given the other issues I identified below, I think it may be best to split off the sink changes and resubmit after the already-approved changes merge.

That is reasonable, I'll back the dd logs changes out of this one 👍

File: website/content/en/highlights/2024-03-26-0-37-0-upgrade-guide.md
Comment (neuronull) [reply website/content/en/highlights/2024-03-26-0-37-0-upgrade-guide.md:1]: good catch, yeah this left over in my editor's cache after I'd moved it and forgot to re-open the correct file to make an addition to it

**Comment by neuronull** [comment]
Okay, I've backed out the dd logs changes on this PR and at this point I think it has the approvals from the agent source changes to get it through.

Now just facing the CI pains...

File: src/sources/datadog_agent/logs.rs
Comment (jszwedko) [inline src/sources/datadog_agent/logs.rs:234]: I might suggest trimming the strings in case they have whitespace like: `env:prod, service:foo`.

File: website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md
Comment (jszwedko) [inline website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md:1]: I was thinking we'd include this change in v0.37.1 along with the sink fixes rather than v0.38.0. Alternatively we put this one in v0.38.0, but we add a fix for v0.37.1 that handles reencoding the tags from the object that they are currently being parsed into when `parse_ddtags` is `true`.

**Comment by jszwedko** [comment]
@neuronull just a note that the `!` should be after the scope for breaking changes. I updated the title.

**Comment by neuronull** [comment]
> @neuronull just a note that the ! should be after the scope for breaking changes. I updated the title.

Thanks!

File: website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md
Comment (neuronull) [reply website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md:1]: Indeed, I was also thinking 0.37.1 . 🤔  I guess I'm a little confused where the upgrade guide note for this needs to go then? I dont think we have these upgrade guides for patch releases right? 😕  I might be crossing wires but I thought we'd discussed needing an upgrade note for this change?

Comment (jszwedko) [reply website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md:1]: It is certainly a breaking change. I would just include it in the changelog entry since, as you note, we don't have a precedent for an upgrade guide for patch releases.

Comment (neuronull) [reply website/content/en/highlights/2024-05-07-0-38-0-upgrade-guide.md:1]: 👍  I'll drop this file

File: src/sinks/datadog/logs/sink.rs
Comment (neuronull) [reply src/sinks/datadog/logs/sink.rs]: In the split off PR for this, I used a different approach, which doesn't require creating this target path for comparison.

File: src/sources/datadog_agent/logs.rs
Comment (jszwedko) [inline src/sources/datadog_agent/logs.rs:231]:
```suggestion
    let ddtags : Vec<Value> = ddtags_str
```

Just noticing that the variable name is a bit misleading now that the value isn't an object.

File: src/sources/datadog_agent/logs.rs
Comment (jszwedko) [inline src/sources/datadog_agent/logs.rs:228]: Nit: is this short-circuiting significant processing? It seems like letting it fall through to the split below would have the same result with minimal overhead.

**Review by jszwedko — APPROVED** [review]
A couple of nits, otherwise 👍

**Pull request merged**
